Add SIP creation panel

Hwesta opened the pull request:
Add the SIP creation panel, which allows arranging files from transfers into SIPs.  Partially described by https://www.archivematica.org/wiki/Transfer_and_SIP_creation#Mockups
- arrange files from Transfers to a new SIP by storing new paths in SIPArrange table.  On create SIP, POST to SS to move from old path to specified path
- SIP arrangement panel
  - left side shows files from backlogged transfers, populated by the backlog search from ES
  - right side shows arranged SIPs, created by creating directories, dragging transfer files over and rearranging them.  Any folder can be selected to create a new SIP
  - logs and metadata cannot be copied over, though they can be viewed
- delete transfers whose contents have been completely stored

Other Features:
- run syncdb by dev-helper
  - to create SIPArrange table
- failed SIP hook
  - sets files in failed SIP available to be arranged again
- backlog now uses SS (artefactual/archivematica-storage-service#9)
- added SIP restructure for files with UUIDs after failing verification, for newly created arranged SIPs that don't have the expected structure
- separation of download from SS vs download from local FS (ie SIP Arrangement)
- cleanup
  - restructure for compliance to files that actually called them
  - required directories and optional files for SIP verification and restructuring moved to common
  - SS url generation centralized to storageService.py helper module
  - remove now unneeded backlog search page - backlog search populates the SIP arrangement panel

Requires artefactual/archivematica-storage-service#9

Still needs more history cleanup.

mcantelon committed:
Re-activated SIP arrange.

mcantelon committed:
Misc tweaks and fixes, added comments

Add directory open/close snapshotting.  Tweaked drag & drop positioning so
it won't mess with hover.  Fixed an issue where elements with spaces in the
CSS ID wouldn't be able to drag and drop. Got rid of spaces in CSS IDs.

mcantelon committed:
Change how files in arrange pane behave

Change file browser so name click handle can be overridden. Change backlog
viewer so you click to highlight a file then click a button to view the
file, rather than clicking on its name. Change so files can be moved within
arrange pane.

mcantelon committed:
Explicit error on abnormal condition.

mcantelon committed:
Roughing out logic to create SIP from arrange files.

mcantelon committed:
Added logging of error.

mcantelon committed:
Removed problem processing XML.

Added logic to rewrite processing config XML to remove pre-selection of
normalization identification method when creating SIPs from the arrange
directory. This is because one method references data which won't exist
because the files haven't actually been ran through the transfer phase
in their current incarnation.

mcantelon committed:
Added UI warning.

Added a pop-up alert if the user is triyng to create
a SIP without first selecting a directory.

mcantelon committed:
Only allow arrange root deletion.

Added UI logic so you can't delete a directory that isn't in
the root level of the arrange directory.

mcantelon committed:
Refactored SIP arrange JS.

Cleaned up SIP arrange JS, fixed minor issue, and moved JS out of
HTML template.

mcantelon committed:
Added copy to arrange security check.

Added logic to prevent users from abusing copy to transfer endpoint.

mcantelon committed:
Fixed minor issues.

Fixed minor issues, including issue with SIP create from backlog.

mcantelon committed:
Work on copy logic.

mcantelon committed:
Refactored, added logic to copy in source METS.

mcantelon committed:
Fixed issue with file copies to arrange.

Fixed an inssue with copying single files into a potential SIP directory
in the SIP arrangement panel.

mcantelon committed:
Added delete from originals dir.

mcantelon committed:
Added paging to SIP browsers.

mcantelon committed:
Added cleanup after delete from backlog.

mcantelon committed:
Minor refactor of open file JS.

Refactor open file JS, add open dir warning, change text from 'open' to
'view'.

mcantelon committed:
Delete backlogged transfers from search page.

mcantelon committed:
Fixed issue with file path padding.

Extensions weren't being preserved which breaks identification.

mcantelon committed:
SIP creation uses default proc. config.

refs #6139

Added logic to add/replace processing configu with default during SIP creation. Tested.

mcantelon committed:
Move SIP creation logic to helpers file.

De-cluttering the ingest views file by moving SIP creation-related
logic into a separate file.

mcantelon committed:
Add ability to create SIP arrange dirs.

refs #6146

Added Django and UI JavaScript logic to allow the creation of
directories in the SIP arrange pane.

mcantelon committed:
Log new transfer file adds in arrange.

refs #6135

Added logging of file/directory copying to a new transfer via the
SIP arrange panel. The log write to logs/arrange.log.

mcantelon committed:
Fixed issues with find by path.

refs #6134

Fixed issues with querying by path in JavaScript file browser.

mcantelon committed:
Added SIP arrange helpers.

mcantelon committed:
Added widget for search results.

refs #6134

Added the EntryList JS class for displayed the results of searches
through file explorers.

mcantelon committed:
Added originals search.

refs #6134

Added way to search originals file browser for specific file/folder
names.

mcantelon committed:
Refined naming, minor fixes.

refs #6134

Refined naming of originals search-related entities and fixed
minor issues.

mcantelon committed:
Can override dir contents path.

refs #6134

Added the ability to choose a different URL path from which to pull
directory contents data. Added this to see if we can pull augmented
directory contents data (for use in file browser search).

mcantelon committed:
Refactoring/cleanup of filesystem ajax.

mcantelon committed:
Added transfer UUID to originals data.

refs #6134

Added the ability to provide additional client-side data from
file structure JSON. Added a specific file structure JSON feed for
the originals file browser that provides the UUID of a transfer
folder. This client-side data can be used for client-side searches.

mcantelon committed:
Added accession ID to originals feed and search

refs #6134

Added accession ID to the originals directory structure feed uses by
the originals file browser. This will allow us to search by accession
ID client-side.  Added client-side accession ID search support to the
originals file browser.

mcantelon committed:
Search result dirs can be opened.

refs #6134

Changed entrylist widget so directories shown in search
results can be opened (and individual files copied from
the result directory).

mcantelon committed:
Refactored toggleDirectory.

refs #6134

mcantelon committed:
Added CSS ID namespaces.

refs #6134

Cleaned up CSS ID creation logic and added ability to provide CSS
ID namespaces to avoid duplicate CSS IDs when a directory entry is
displayed more than once in an entry list.

mcantelon committed:
Fixed JS/CSS typos.

refs #6134

mcantelon committed:
Removed hardcoding of template.

refs #6134

Removed hardcoding of template in EntryList JS class.

mcantelon committed:
Added paging to search results.

refs #6134

Added paging to the SIP arrangement originals search.

mcantelon committed:
Refactor page link rendering.

refs #6134

Refactored paging link rendering, of directory children in the file
explorer, to use the same rendering logic as used in the entry
list.

mcantelon committed:
Added "no results" display.

refs #6134

In the SIP arrangement originals search, added "No results" display
if no search results are found.

mcantelon committed:
Minor originals search fix.

refs #6134

Fixed a minor issue wth the originals search: made it so only
the top-level node's children show up as search results.

mcantelon committed:
Got rid of inline styling.

refs #6134

Replaced inline CSS styling with use of classes in file explorer paging links.

mcantelon committed:
Initial JSdoc documentation.

refs #6134

Started work using JSdoc to document the file explorer.

mcantelon committed:
Fixing minor issue.

Hwesta committed:
Update pruning processing config

Fix problem where file isn't actually written to disk after
modification (possibly because not closed?).  Ensure file is closed after
use.

mcantelon committed:
AJAX-related issue with snapshotting.

refs #6134

Added the ability for file explorer open/close state to be preserved
after AJAX request when entries are configured to read children via
AJAX.

mcantelon committed:
File browser handles AJAX child URLs.

refs #6134

The file browser extends the file explorer. Added the
ability to create a file browser specifying an AJAX child URL
and added different refreshing behavior if that is set.

mcantelon committed:
Scalability improvement to SIP create browsers.

refs #6134

Change the originals and arrange file browsers to
pull directory data via per-directory requests rather
than pulling the entire directory subtrees at once.

Hwesta committed:
Cleanup

Fix headers - remove duplicates and reorder to reduce future duplicates.
Remove unused model.

Hwesta committed:
Add SIP Arrangement table

refs #6131

Add table to track SIP Arrangement in the database.  Schema still flexible.

Hwesta committed:
Add dashboard syncdb to dev-helper script

Update recreateDB script called by dev-helper to run syncdb for the
dashboard.  Add new permissions for archivematica to run syncdb properly
(CREATE, ALTER, INDEX).

Hwesta committed:
Set up dashboard testing fixtures

Add config for testing fixtures folder, and an initial test user fixture.

Hwesta committed:
Add arrange contents view that reads from DB

refs #6131

Add view that queries SIPArrange table and returns the entries and
directories for Javascript file-browser.  Add test for that function.

Hwesta committed:
Add arrange contents view that reads from DB

refs #6131

Add view that queries SIPArrange table and returns the entries and
directories for Javascript file-browser.  Add test for that function.

Hwesta committed:
Update move with arrange functionality

refs #6131

Update move_within_arrange endpoint to expect the SIPArrange table and
adjust paths within that tables as needed.  Refactor moveHandler to have
fewer nested ifs, send correct information for SIPArrange table.

Hwesta committed:
Require arrange paths to be inside /arrange/

refs #6131

Require arrange paths in SIPArrange to start with DEFAULT_ARRANGE_PATH, and
enforce that for arrange_contents and move_within_arrange.

Hwesta committed:
Update originals view

refs #6131

Update where the originals pane gets its data from.  Archivematica handles
querying the SS for the correct location.

Hwesta committed:
Remove unused functions

Remove copy_to_originals, helper function, and unused transfer browser.
Transfer browser was an early attempt at SIP arrangement, superseded by
existing SIP arrangement.

Hwesta committed:
Update copy to arrange functionality

refs #6131

Update copy_to_arrange endpoint to expect the SIPArrange table.
Instead of moving files on disk, add entries to the SIPArrange table
with the original path and the new arranged path.  Since only the
source and destination path are sent, fetch the full directory
structure from backlog Location.

Hwesta committed:
Update directory handling in arrange panel

refs #6131

Update create directory function to expect SIPArrange table.
Directories in SIPArrange now have NULL for the source path and
file_uuid, as there may be no original source, and we only care about
moving the actual files around.  Updated unit test and fixture to match.

Hwesta committed:
Update delete from arrange functionality

refs #6131

Update delete endpoint to check if a file that doesn't exist is part of
SIP arrangement and delete it from the SIPArrange table.

Hwesta committed:
Misc fixes

refs #6131

Remove file UUIDs for directories, and added file UUID to SIPArrange
creation.  Remove passing shared directory information to JS, as that
will be handled Django-side.  Remove originals panel search, and moved
transfer backlog search to the top of the page.

Hwesta committed:
Update SIPArrange model

refs #6131

Hwesta committed:
Update information indexed for Transfers

refs #6131

All files in a Transfer will be indexed so that the metadata and logs will
be available to be viewed in the SIP Arrangement panel.  Indexing and
updating the index on a file do not verify that they are in the DB anymore.
The full file path is not indexed, but the relative path (including the
transfer folder) is.

Add a wrapper for search_raw that queries for MAX_QUERY_SIZE entries from
ElasticSearch.  The default number returned is 10, and there is no way to
request all entries that match a search.

Hwesta committed:
Backlog search populates originals panel

refs #6131

Several updates to make the transfer backlog search populate the originals
panel in SIP Arrangement.

Backlog.js queries the dashboard, which queries ElasticSearch based on the
provided query parameters.  The resulting paths are rearranged into a format
that the Javascript file-explorer is expecting and returned.  Backlog.js
updates the originals panel and forces a re-render.

Cleanup of unneeded functions: originals content endpoint, create/delete JS
from backlog.js, backlog viewing template, remainder of originals pane
search code, misc other helper functions.

Hwesta committed:
Update delete and hide buttons

refs #6131

Originals hide button now hides the selected folder and all its children.
Arrange delete button given its own endpoint, so we don't have to guess
when a delete is for arrange instead of the filesystem.  file_browser
delete URL made configurable to support this.

Hwesta committed:
Open backlog folder when populated with data

refs #6131

Hwesta committed:
Update move from originals to arrange

refs #6131

If the source folder is the objects directory, add the contents of the
objects directory, not the directory itself.  Rearranged copy_to_arrange to
accomodate change and error checking.

Create sip has to be top level directory (cannot be a subfolder or the
arrange directory).

Hwesta committed:
Minor fixes

refs #6131

Delete doesn't delete directories that start with the same characters.
Create directory can create top level directories if nothing is selected.
Originals panel has some default text.  Don't display the 'files?' option
in backlog search.

Hwesta committed:
Logs and metadata cannot be dragged

refs #6131

Add attribute to transfer backlog results that specifies logs and metadata
directories (and their children) as not draggable.  Update the file browser
to recognize that attribute, and disable the drag & drop handler for those
elements.  Style those elements grey.

Hwesta committed:
Save file UUID with arranged files

refs #6131

Fetch file UUID from ElasticSearch for arranged files, based on the original
relative path. Strip the UUID from a transfer name when creating a SIP from
a Transfer.

Hwesta committed:
Don't enforce unique arranged paths

refs #6131

Enforcing unique arranged paths causes problems when a SIP that has been
created had the same path.  Remove requirement for unique arranged paths
for now.

Hwesta committed:
Logs and metadata added to SIP

refs #6131

Logs and metadata folders added to SIP when created, or empty folders
created if they don't exist.

Hwesta committed:
Arranged files are not draggable

refs #6131

A file should not be multiple SIPs.  If a file has been arranged (can be
found in SIPArrange.original_files), then it is greyed out in the originals
panel. When a file is dragged to the arranged panel, or deleted from the
arranged panel, the originals panel refreshes to get the updated status.

SIPArrange original paths are now unique, and when copying a folder, does
not copy files from originals that have already been arranged.

Hwesta committed:
Update for base64 path encoding

refs #6067, #6131

All paths sent via JSON are encoded in base64, to handle cases where
they contain non-unicode characters.  Files in SIP arrangement should
all be in unicode, but because it shares code with transfer source, the
code should be able to handle non-unicode characters, and be internally
consistent.

filesystem_ajax and ingest views updated to expect base64 paths and
return base64 paths as needed.  file-explorer and file-browser updated
to expect and return base64 paths as well.  browse_location encodes
outgoing paths to base64, and decodes the returned entries.

Also, other fixes from rebase.

Hwesta committed:
Move restructureForCompliance functions to correct client script

restructureForCompliance.py contained several functions that were not
used in that script, but were imported into other client scripts.  Move
those functions to the correct client script.  Add common code to
archivematicaFunctions and import that.

Hwesta committed:
Restructure if verify SIP compliance fails

refs #6131

Add microservice to restructure a SIP for compliance, where the files
are already stored in the DB.  Insert microservice after verify SIP for
compliance, if it fails.

Hwesta committed:
Move to backlog uses storage service

refs #6131

Add moveToBacklog client script that uses the storage service to move the
Transfer to the configured backlog Location.  Update storeAIP to send
relative AIP path instead of absolute.  Update output when moving a file.

Hwesta committed:
Update backlog files' status

refs #6131

Update arranged status of files to track when a SIP fails or completes
successfully.

Index AIP marks files an being in an AIP, and submits a
deletion request to the storage service when all files in a transfer
are in an AIP.  Transfer UUID is tracked for each arranged file to
enable this.

Failed and rejected SIPs run failedSIPCleanup.py, which deletes
SIPArrange rows for each file in the SIP, making them arrangeable again.

Hwesta committed:
Add view file to SIP arrangement

refs #6131

Add two download files to filesystem_ajax: one that proxies to the
storage service, another which reads from local disk. Update view file
to use the storage servie, update browse AIP to use the local FS.
Update archival storage to use storage service helpers, so the base
storage service URL is only fetched in one place.

Hwesta committed:
Fixes for ElasticSearch update

refs #6131

ElasticSearch 0.90/Lucene 4 now interprets / as being for regex, and they
must be escaped.

Hwesta committed:
Allow SIPs to be created from any level

refs #6131

SIPs can be created from a folder at any level.  Change how the new SIP
name is parsed out, and add extra error checking to ensure that a SIP
cannot be created from a file.

Hwesta committed:
Move logs/metadata to completedTransfers

refs #6131

Later microservices copy metadata and logs from the completedTransfers
directory, so copying the expected files there.

Hwesta committed:
First pass at creating a SIP

refs #6131

Create SIP button in SIP Arrangement panel moves arranged files from backlog
to the processing space.  Database information not set up yet, as that needs
to be fetched from ElasticSearch or the Storage Service.

SIPArrange entries that are folders no longer have a source path, as this
leads to confusion when moving files.  arrange_paths are now unique, since
an arranged file shouldn't have multiple sources.

Add new create SIP function for arranged SIPs, and remove the old create
SIP function, helper functions, and endpoint.  Create arranged SIP
function creates the SIP object, adds the arranged files to that SIP and
updates the locations.  WARNING this only works with files that
originated on this pipeline.  Also, clean up headers.

mistydemeo commented:
You already know this, so just marking down so I remember: the failed SIP hook should probably take https://projects.artefactual.com/issues/6636 into account. I imagine weird things would happen if a rejected DIP ends up being slurped back as a SIP for arrangement?

mistydemeo commented:
> backlog now uses SS (artefactual/archivematica-storage-service#9)

Yay! More and more stuff that should be using the SS is using it :D

mistydemeo commented on src/MCPClient/lib/clientScripts/failedSIPCleanup.py:
```diff
@@ -0,0 +1,29 @@
+import sys
+
+path = '/usr/share/archivematica/dashboard'
+if path not in sys.path:
```
I've seen this pattern a few times: just out of curiosity, do we anticipate these client scripts being imported by something that is already using dashboard paths?

Hwesta replied:
I noticed that the sys.path is often cluttered with multiple instances of '/usr/share/archivematica/dashboard', '/usr/lib/archivematica/archivematicaCommon' etc having been added, and was hoping to prevent that (what happens if sys.path gets absurdly large?).  It seems less useful in client scripts as they're probably run once and don't have this problem. I go back and forth on whether it's a worthwhile pattern to cart around.

mistydemeo commented on src/MCPClient/lib/clientScripts/indexAIP.py:
```diff
@@ -10,13 +10,14 @@
 import databaseInterface
 import elasticSearchFunctions
 from executeOrRunSubProcess import executeOrRun
+import storageService as storage_service
```
Outside the scope of this PR, but let's rename the module at some point!

Hwesta replied:
I originally named it storageService to match with everything else in archivematicaCommon, but import it as storage_service to make myself happy. :P PEP8 is conflicting - recommends pothole, but also says go with surrounding code style.  Would like to rename though!

mistydemeo commented on src/MCPClient/lib/clientScripts/moveToBacklog.py:
```diff
@@ -0,0 +1,66 @@
+
+
+def main(transfer_uuid, transfer_path):
+    current_location = storage_service.get_location(purpose="CP")[0]
```
Do we have a constant for this purpose?

Hwesta replied:
Yes, but it's in the storage service and not import-able here. :( Should I redefine some SS constants here?  In archivematicaCommon/storageService? Is there a better solution?

mistydemeo commented on src/MCPClient/lib/clientScripts/restructureForCompliance.py:
```diff
@@ -23,130 +23,25 @@
-                fileUUID = value
-                updateFileGrpUse(fileUUID, "TRIM metadata")
 
-def restructureBagForComplianceFileUUIDsAssigned(unitPath, unitIdentifier, unitIdentifierType, unitPathReplaceWith = "%transferDirectory%"):
```
:D :D :D

mistydemeo commented on src/archivematicaCommon/lib/archivematicaFunctions.py:
```diff
@@ -23,9 +23,20 @@
-import sys
 import re
 
+REQUIRED_DIRECTORIES = [
```
:thumbsup:

mistydemeo commented on src/dashboard/src/components/filesystem_ajax/helpers.py:
```diff
@@ -0,0 +1,106 @@
+    # return fully traversed data
+    return directory
+
+def directory_contents(path, contents=[]):
```
IIRC I deleted this in another branch, possibly in qa/1.x - just FYI in case you get a merge conflict from it.

Hwesta replied:
Thanks, deleted directory_contents in my branch as well.

mistydemeo commented on src/dashboard/src/components/filesystem_ajax/views.py:
```diff
@@ -342,96 +283,382 @@ def copy_to_start_transfer(request):
+    sip_config_path = os.path.join(staging_abs_path, 'processingMCP.xml')
+    shutil.copyfile(default_config_path, sip_config_path)
+
+    # # Remove pre-normalize file format ID, may not have been ID'd yet
```
Planning to keep or remove this?

Hwesta replied:
Removed it, since I don't think the processingConfig changes are needed after all, as SIP
creation requires the transfers to be on the same pipeline as the new SIP currently. That will eventually be replaced by parsing a transfer METS file into the DB, so all the file  information and events will be there, and file ID won't have to be redone.

mistydemeo commented on src/dashboard/src/main/models.py:
```diff
@@ -227,6 +227,25 @@ class Transfer(models.Model):
 
+
+class SIPArrange(models.Model):
+    # original_location = models.CharField(max_length=36)
```
Remove the commented out line?

mistydemeo replied:
So this model represents a file in an arranged SIP?

mistydemeo commented on src/dashboard/tests/test_filesystem_ajax.py:
```diff
@@ -0,0 +1,54 @@
+#! /usr/bin/python2
```
Yay, tests! :sparkling_heart:

mistydemeo commented:
This looks good! I admit I skimmed the JS, but otherwise looks like it's in good shape

mistydemeo commented:
Can the same file(s) appear in multiple SIPs?

Hwesta commented:
Updated failedSIPCleanup to run before the (likely to fail) move to failed directory.  Since the post-failed SIP hook only touches the SIPArrange table, DIPs should be unaffected.

The same file should not appear in multiple SIPs. Any entry in SIPArrange with the same original_path greys out the corresponding entry in the originals panel.  Once an AIP is created, store_aip checks to see if all files in a backlogged transfer are in AIPs (as tracked by the SIPArrange table), and submits a delete request if that's the case.  It's possible to get around these restrictions, but casual error should be avoided like this.

Hwesta commented:
Rebased on current qa/1.x, with updates mentioned above.

Hwesta committed:
Update & add more tests

refs #6131

Update tests for code changes, including base64 and a default arrange
directory.  Add tests for new functions.  Add new more valid fixture
data.
